### PR TITLE
Initial reaction sdk (nodejs)

### PR DESCRIPTION
**Pull request by danielgerlag**
# Description

- Added TypeSpec definitions for contract between query and reaction.
- TypeSpec generates the models for each specific language SDK
- Added `externalImage` flag to services for sources/reaction to indicate that an image should/not be pulled from the Drasi registry.  This enables users to publish their reaction/source images to their own registry.
- Created initial NodeJS SDK for Reactions, along with samples.

---
# Reaction SDK for Drasi

This library provides the building blocks and infrastructure to implement a [Drasi](https://drasi.io/) Reaction in Node.js

### Install the package

```
npm install --save @drasi/reaction-sdk
```

### Basic example

The following example simply breaks down and logs the various parts of the incoming change event from a [Continuous Query](https://drasi.io/concepts/continuous-queries/).

```typescript
import { DrasiReaction, ChangeEvent } from '@drasi/reaction-sdk';

let myReaction = new DrasiReaction(async (event: ChangeEvent) => {

    console.log(`Received change sequence: ${event.sequence} for query ${event.queryId}`);

    for (let added of event.addedResults) {
        console.log(`Added result: ${JSON.stringify(added)}`);
    }

    for (let deleted of event.deletedResults) {
        console.log(`Removed result: ${JSON.stringify(deleted)}`);
    }

    for (let updated of event.updatedResults) {
        console.log(`Updated result - before: ${JSON.stringify(updated.before)}, after: ${JSON.stringify(updated.after)}`);
    }
});

myReaction.start();
```

### A more advanced example

The following example illustrates 
 - Retrieving a configuration value from the Reaction manifest
 - Defining a custom per query configuration object
 - Parsing the per query configuration object from Yaml
 - Process change events form the query
 - Process control events (start, stop, etc.) from the query

```typescript
import { DrasiReaction, ChangeEvent,parseYaml, ControlEvent, getConfigValue } from '@drasi/reaction-sdk';

// Retrieve the connection string from the Reaction configuration
const myConnectionString = getConfigValue("MyConnectionString");

// Define a custom per query configuration object
class MyQueryConfig {
    greeting: string = "Default greeting";
}

// Define the function that will be called when a change event is received
async function onChangeEvent(event: ChangeEvent, queryConfig?: MyQueryConfig): Promise<void> {
    // We can access the per query config here
    console.log(queryConfig.greeting);  

    // do something with the result set diff
}

// Define the function that will be called when a control event is received
async function onControlEvent(event: ControlEvent, queryConfig?: MyQueryConfig): Promise<void> {    
    console.log(`Received control signal: ${JSON.stringify(event.controlSignal)} for query ${event.queryId}`);    
}

console.log(`Starting Drasi reaction with connection string: ${myConnectionString}`);

// Configure the Reaction with the onChangeEvent and onControlEvent functions
let myReaction = new DrasiReaction<MyQueryConfig>(onChangeEvent, {
    parseQueryConfig: parseYaml,  // Parse the per query configuration from Yaml
    onControlEvent: onControlEvent
});

// Start the Reaction
myReaction.start();
```

### CODE REVIEW

**Comment by ruokun-niu** [comment]
Seems like we are able to retrieve the queyrId from the event 
```js
console.log(`Received change sequence: ${event.sequence} for query ${event.queryId}`);
```

Does this mean that we do not need to specify the queryIds in the reaction yaml file anymore?

File: control-planes/mgmt_api/src/api/v1/models/providers.rs [hunk 1]
```diff
@@ -28,6 +28,8 @@ pub struct ProviderSpecDto {
 #[derive(Serialize, Deserialize, Debug)]
 pub struct ProviderServiceDto {
     pub image: String,
+    #[serde(rename = "externalImage")]
+    pub external_image: Option<bool>,
```
Comment (ruokun-niu) [inline control-planes/mgmt_api/src/api/v1/models/providers.rs:32]: What does external_image do?

**Comment by danielgerlag** [comment]
> Seems like we are able to retrieve the queyrId from the event
> 
> ```js
> console.log(`Received change sequence: ${event.sequence} for query ${event.queryId}`);
> ```
> 
> Does this mean that we do not need to specify the queryIds in the reaction yaml file anymore?

We still do, because you might not want a reaction to listen to all queries. You will only get events for the queries configured in the yaml, but this tells you which one of those it is.

File: control-planes/mgmt_api/src/api/v1/models/providers.rs (hunk 1 above)
Comment (danielgerlag) [reply control-planes/mgmt_api/src/api/v1/models/providers.rs:32]: > - Added externalImage flag to services for sources/reaction to indicate that an image should/not be pulled from the Drasi registry. This enables users to publish their reaction/source images to their own registry.

**Comment by ruokun-niu** [comment]
> > Seems like we are able to retrieve the queyrId from the event
> > ```js
> > console.log(`Received change sequence: ${event.sequence} for query ${event.queryId}`);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Does this mean that we do not need to specify the queryIds in the reaction yaml file anymore?
> 
> We still do, because you might not want a reaction to listen to all queries. You will only get events for the queries configured in the yaml, but this tells you which one of those it is.

Ah I see, is this how the Reaction will know which queries it should listen for (copied from the advanced example)?

```js
let myReaction = new DrasiReaction(onChangeEvent, {
    parseQueryConfig: parseYaml, // Parse the per query configuration from Yaml
    onControlEvent: onControlEvent
});
```

**Review by ruokun-niu — APPROVED** [review]

**Comment by danielgerlag** [comment]
> > > Seems like we are able to retrieve the queyrId from the event
> > > ```js
> > > console.log(`Received change sequence: ${event.sequence} for query ${event.queryId}`);
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Does this mean that we do not need to specify the queryIds in the reaction yaml file anymore?
> > 
> > 
> > We still do, because you might not want a reaction to listen to all queries. You will only get events for the queries configured in the yaml, but this tells you which one of those it is.
> 
> Ah I see, is this how the Reaction will know which queries it should listen for from the yaml(copied from the advanced example)?
> 
> ```js
> let myReaction = new DrasiReaction(onChangeEvent, {
>     parseQueryConfig: parseYaml, // Parse the per query configuration from Yaml
>     onControlEvent: onControlEvent
> });
> ```

No, that is how it extracts per query config. The list of queries to listen to is dealt with internally by the library, the user does not need to think about it.

**Comment by ruokun-niu** [comment]
> > > > Seems like we are able to retrieve the queyrId from the event
> 
> > > > ```js
> 
> > > > console.log(`Received change sequence: ${event.sequence} for query ${event.queryId}`);
> 
> > > > ```
> 
> > > > 
> 
> > > > 
> 
> > > >     
> 
> > > >       
> 
> > > >     
> 
> > > > 
> 
> > > >       
> 
> > > >     
> 
> > > > 
> 
> > > >     
> 
> > > >   
> 
> > > > Does this mean that we do not need to specify the queryIds in the reaction yaml file anymore?
> 
> > > 
> 
> > > 
> 
> > > We still do, because you might not want a reaction to listen to all queries. You will only get events for the queries configured in the yaml, but this tells you which one of those it is.
> 
> > 
> 
> > Ah I see, is this how the Reaction will know which queries it should listen for from the yaml(copied from the advanced example)?
> 
> > 
> 
> > ```js
> 
> > let myReaction = new DrasiReaction(onChangeEvent, {
> 
> >     parseQueryConfig: parseYaml, // Parse the per query configuration from Yaml
> 
> >     onControlEvent: onControlEvent
> 
> > });
> 
> > ```
> 
> 
> 
> No, that is how it extracts per query config. The list of queries to listen to is dealt with internally by the library, the user does not need to think about it.

Oh that's cool. I'm so not used to the fact that we don't need to explicitly pubsub  the queries

**Review by agentofreality — APPROVED** [review]

**Pull request merged**